File: matrixproduct.py
from time import perf_counter
import sys
'''
from pypapi import papi_low as papi
from pypapi import events
'''

file = None


def main(): 
    '''
    papi.library_init()
    evs = papi.create_eventset()
    papi.add_event(evs,events.PAPI_L1_DCM) 
    papi.add_event(evs,events.PAPI_L2_DCM)
    '''

    args = sys.argv[1:]

    if args and args[0] == "test":
        handleTestCases()
        return
    
    while True:
        print("1. Matrix Product")
        print("2. Line Multiplication")
        print("3. Block Multiplicaton")
        print("4. Close Program")
        option = int(input("Select an option: "))

        if(option == 4):
            break

        # Number of lines and columns
        n = int(input("Choose Matrix Dimension:\nLines x columns: "))
        
        '''
        papi.start(evs)
        '''

        match option:
            case 1:
                matrix_product(n)
            case 2:
                line_multiplication(n)
            case 3:
                block_size = int(input("Choose Block Size: "))
                block_multiplication(n,block_size)
                continue
            case 4:
                break
            case _:
                print("Invalid option, try again")
                continue
        '''
        result = papi.stop(evs)
        print(result)   
        '''
    '''
    papi.cleanup_eventset(evs)
    papi.destroy_eventset(evs)   
    '''

def handleBlockFunctions(n,block_size):
    print("1. Normal Block Matrix Multiplication") 
    print("2. Block Matrix Multiplication with Inline Multiplication")
    option = int(input("Choose an option: "))
    match option:
        case 1:
            block_multiplication(n,block_size)
        case 2:
            block_line_multiplication(n,block_size)
        case _:
            print("Invalid input")

def matrix_product(n):
    matrix1 = []
    matrix2 = []
    result = []

    for _ in range(n):
        for _ in range(n):
            matrix1.append(1.0) 

    for i in range(n): 
        for _ in range(n):
            matrix2.append(i+1)  

    for _ in range(n): 
        for _ in range(n): 
            result.append(0)

    # Put clock start here 
    time_start = perf_counter() 

    for i in range(n): 
        for j in range(n): 
            temp = 0
            for k in range(n): 
                temp += matrix1[i*n + k] * matrix2[k*n + j]
            result[i*n + j] = temp 

    # Put clock end here 
    time_finish = perf_counter()

    print(f"Time: {time_finish - time_start} seconds")

    print("Result Matrix:")
    for j in range(min(10, n)):
        print(f"{result[j]} ", end='')
    print("\n")
    

def line_multiplication(n):
    matrix1 = []
    matrix2 = []
    result = []

    for _ in range(n):
        for _ in range(n):
            matrix1.append(1.0) 

    for i in range(n): 
        for _ in range(n):
            matrix2.append(i+1)  

    for _ in range(n): 
        for _ in range(n): 
            result.append(0)

    # Put clock start here 
    time_start = perf_counter() 

    for i in range(n): 
        for j in range(n):
            temp =  matrix1[j + i*n]
            for k in range(n):
                result[k + i*n]+= temp*matrix2[k + j*n] # matrix1[j + i*n] caso substituir pelo temp mais rápido, devido ao valor cached
                

    # Put clock end here 
    time_finish = perf_counter()

    print(f"Time: {time_finish - time_start} seconds")

    print("Result Matrix:")
    for j in range(min(10, n)):
        print(f"{result[j]} ", end='')
    print("\n")

# Version with line block multiplication
def block_line_multiplication(n,block_size):
    matrix1 = []
    matrix2 = []
    result = []

    for _ in range(n):
        for _ in range(n):
            matrix1.append(1.0) 

    for i in range(n): 
        for _ in range(n):
            matrix2.append(i+1)  

    for _ in range(n): 
        for _ in range(n): 
            result.append(0)

    # Put clock start here 
    time_start = perf_counter()

    for bi in range(0, n, block_size):  
        for bj in range(0, n, block_size):  
            for bk in range(0, n, block_size):  
                for i in range(bi, min(bi + block_size, n)): 
                    for j in range(bj, min(bj + block_size, n)):
                        temp =  matrix1[j + i*n]
                        for k in range(bk, min(bk + block_size, n)):
                            result[k + i*n]+= temp*matrix2[k + j*n] # matrix1[j + i*n] caso substituir pelo temp mais rápido, devido ao valor cached
    # Put clock end here 
    time_finish = perf_counter()

    print(f"Time: {time_finish - time_start} seconds")

    print("Result Matrix:")
    for j in range(min(10, n)):
        print(f"{result[j]} ", end='')
    print("\n")


def block_multiplication(n,block_size):
    matrix1 = []
    matrix2 = []
    result = []

    for _ in range(n):
        for _ in range(n):
            matrix1.append(1.0) 

    for i in range(n): 
        for _ in range(n):
            matrix2.append(i+1)  

    for _ in range(n): 
        for _ in range(n): 
            result.append(0)

    # Put clock start here 
    time_start = perf_counter() 

    for bi in range(0, n, block_size):  
        for bj in range(0, n, block_size): 
            for bk in range(0, n, block_size):  
                
                minValueI = min(bi + block_size, n)
                minValueJ = min(bj + block_size, n)
                minValueK = min(bk + block_size, n)
                for i in range(bi, minValueI):
                    for j in range(bj, minValueJ):
                        sum_value = 0  
                        for k in range(bk, minValueK):
                            sum_value += matrix1[i * n + k] * matrix2[k * n + j]
                        result[i * n + j] += sum_value

    # Put clock end here 
    time_finish = perf_counter()

    print(f"Time: {time_finish - time_start} seconds")

    print("Result Matrix:")
    for j in range(min(10, n)):
        print(f"{result[j]} ", end='')
    print("\n")


def handleTestCases():

    file = open("docs/data_py.csv", "w")

    file.write("functionType,L2 DCM,L1 DCM,MatrixSize,BlockSize,NumThreads,Real Time")

    print("== Normal multiplication tests ==", end='')

    for n in range(600, 3001, 400):
        testFunc(matrix_product, n)
    
    print(" Complete!")

    print("== Line multiplication tests ==",end='')
    for n in range(600, 3001, 400):
        testFunc(line_multiplication, n)
    
    print(" Complete!")


    print("== Normal Block multiplication tests ==",end='')

    for n in range(4096, 10241,2048):
        bkSizes = [128,256,512]
        for bkSize in bkSizes:
            testBlockFunc(block_multiplication,n,bkSize)

    print(" Complete!")


    print("== Block multiplication with Inline Multiplication tests ==",end='')
    for n in range(4096, 10241,2048):
        bkSizes = [128,256,512]
        for bkSize in bkSizes:
            testBlockFunc(block_line_multiplication,n,bkSize)
            
    print(" Complete!")

    file.close()



def testFunc( f, n):
    avg = 0.0

    for i in range(3):
        avg += f(n)
    
    avg /= 3.0

    writeToCsv(f.__name__,avg,n)

def testBlockFunc( f , n , bkSize):
    avg = 0.0

    for i in range(3):
        avg += f(n,bkSize)
    
    avg /= 3.0

    writeToCsv(f.__name__,avg,n,bkSize)


def writeToCsv(funcName, time, n, bkSize=None): 
    if funcName == "matrix_product":
        csvName = "Normal Mult"
    elif funcName == "line_multiplication":
        csvName = "Inline Mult"
    elif funcName == "block_multiplication":
        csvName = "Block Mult"
    elif funcName == "block_line_multiplication":
        csvName = "Inline Block Mult"
    else:
        raise ValueError("Incorrect function name")
    
    

    if bkSize == None:
        csvBkSize = -1
    else:
        csvBkSize = bkSize


    file.write(f"{csvName},-1,-1,{n},{csvBkSize},-1,{time}\n")





if __name__ == "__main__":
    main()

File: src/go/matrixproduct.go
package main

import (
	"fmt"
	"time"
	"os"
)

func OnMult(n int) float64{
	matrixA := make([]float64, n*n)
	matrixB := make([]float64, n*n)
	matrixC := make([]float64, n*n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixA[i*n+j] = 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixB[i*n+j] = float64(i) + 1.0
		}
	}



	start := time.Now()

	var temp float64
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			temp = 0
			for k := 0; k < n; k++ {
				temp += float64(matrixA[i*n+k] * matrixB[k*n+j])
			}
			matrixC[i*n+j] = temp

		}
	}

	end := time.Now()

	elapsed := end.Sub(start).Seconds()

	/*
	fmt.Println("Time: ", elapsed, " seconds")
	fmt.Println("Result matrix: ")
	for i := 0; i < 1; i++ {
		for j := 0; j < min(10, n); j++ {
			fmt.Print(matrixC[j], " ")
		}
	}
	fmt.Println()
	*/
	return elapsed

}

func OnMultLine(n int) float64 {
	matrixA := make([]float64, n*n)
	matrixB := make([]float64, n*n)
	matrixC := make([]float64, n*n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixA[i*n+j] = 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixB[i*n+j] = float64(i) + 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixC[i*n+j] = 0
		}
	}

	start := time.Now()
	var temp float64
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			temp = matrixA[i*n+j]
			for k := 0; k < n; k++ {
				matrixC[i*n+k] += temp * matrixB[j*n+k]
			}
		}
	}

	end := time.Now()

	elapsed := end.Sub(start).Seconds()
	/*
	fmt.Println("Time: ", elapsed, " seconds")
	fmt.Println("Result matrix: ")
	for i := 0; i < 1; i++ {
		for j := 0; j < min(10, n); j++ {
			fmt.Print(matrixC[j], " ")
		}
	}
	fmt.Println()
	*/
	return elapsed
}

func OnMultBlockLine(n int, bkSize int) float64 {
	matrixA := make([]float64, n*n)
	matrixB := make([]float64, n*n)
	matrixC := make([]float64, n*n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixA[i*n+j] = 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixB[i*n+j] = float64(i) + 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixC[i*n+j] = 0
		}
	}

	start := time.Now()

	for bi := 0; bi < n; bi += bkSize {
		for bj := 0; bj < n; bj += bkSize {
			for bk := 0; bk < n; bk += bkSize {
				minValueI := min(bi+bkSize, n)
				minValueJ := min(bj+bkSize, n)
				minValueK := min(bk+bkSize, n)

				for i := bi; i < minValueI; i++ {
					for j := bj; j < minValueJ; j++ {
						sum_value := matrixA[j+i*n]
						for k := bk; k < minValueK; k++ {
							matrixC[k+i*n] += sum_value * matrixB[k+j*n]
						}
					}
				}
			}
		}
	}

	end := time.Now()

	elapsed := end.Sub(start).Seconds()

	fmt.Println("Time: ", elapsed, " seconds")
	fmt.Println("Result matrix: ")
	for i := 0; i < 1; i++ {
		for j := 0; j < min(10, n); j++ {
			fmt.Print(matrixC[j], " ")
		}
	}
	fmt.Println()

	return elapsed
}

func OnMultBlock(n, bkSize int) float64 {
	matrixA := make([]float64, n*n)
	matrixB := make([]float64, n*n)
	matrixC := make([]float64, n*n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixA[i*n+j] = 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixB[i*n+j] = float64(i) + 1.0
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			matrixC[i*n+j] = 0
		}
	}

	start := time.Now()
	var temp float64
	for bi := 0; bi < n; bi += bkSize {
		for bj := 0; bj < n; bj += bkSize {
			for bk := 0; bk < n; bk += bkSize {
				minValueI := min(bi+bkSize, n)
				minValueJ := min(bj+bkSize, n)
				minValueK := min(bk+bkSize, n)

				for i := bi; i < minValueI; i++ {
					for j := bj; j < minValueJ; j++ {
						temp = 0
						for k := bk; k < minValueK; k++ {
							temp += matrixA[i*n+k] * matrixB[k*n+j]
						}
						matrixC[i*n+j] += temp
					}
				}
			}
		}
	}

	end := time.Now()

	elapsed := end.Sub(start).Seconds()

	fmt.Println("Time: ", elapsed, " seconds")
	fmt.Println("Result matrix: ")
	for i := 0; i < 1; i++ {
		for j := 0; j < min(10, n); j++ {
			fmt.Print(matrixC[j], " ")
		}
	}
	fmt.Println()

	return elapsed
}

func chooseBlockFunction(n int, blockSize int) {
	var op int
	fmt.Println("1. Normal Block Matrix Multiplication")
	fmt.Println("2. Block Matrix Multiplication with Inline Multiplication")
	fmt.Print("Selection?: ")
	fmt.Scan(&op)

	switch op {
	case 1:
		OnMultBlock(n, blockSize)
	case 2:
		OnMultBlockLine(n, blockSize)
	default:
		fmt.Println("Invalid Input")
	}
}

func main() {

	args := os.Args[1:]
	var op int
	var dimensions int
	var bkSize int

	if( len(args) > 0 && args[0] == "test"){
		handleTestCases();
	}

	for {
		fmt.Println("1. Multiplication")
		fmt.Println("2. Line Multiplication")
		fmt.Println("3. Block Multiplication")
		fmt.Println("0. Exit Program")
		fmt.Print("Selection?: ")
		fmt.Scan(&op)

		if op == 0 {
			break
		}

		fmt.Print("Dimension: lins=cols ? ")
		fmt.Scan(&dimensions)

		switch op {
		case 1:
			OnMult(dimensions)
		case 2:
			OnMultLine(dimensions)
		case 3:
			fmt.Print("Block Size? ")
			fmt.Scan(&bkSize)
			chooseBlockFunction(dimensions, bkSize)
		}

	}
}

func handleTestCases() {

	file, err := os.OpenFile("../docs/data_go.csv", os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer file.Close()

	
	fmt.Print("== Normal multiplication tests ==")
	for n:= 600; n <= 3000; n += 400{
		testFunc(file, OnMult,"Normal Mult",n)
	}
	fmt.Println(" Complete!")
	
	
	fmt.Print("== Line multiplication tests ==")
	for n := 600; n <= 3000; n+= 400{
		testFunc(file, OnMultLine,"Inline Mult",n)
	}
	fmt.Println(" Complete!")
	

	/*
	fmt.Print("== Normal Block multiplication tests ==")
	for n := 4096; n <= 10240; n+= 2048{
		for bkSize := 128; bkSize <= 512; bkSize*=2{
			testBlockFunc(OnMultBlock,n,bkSize)
		}
	}
	fmt.Println(" Complete!")

	fmt.Print("== Block multiplication with Inline Multiplication tests ==")
	for n := 4096; n <= 10240; n+= 2048{
		for bkSize := 128; bkSize <= 512; bkSize*=2{
			testBlockFunc(OnMultBlockLine,n,bkSize)
		}
	}
	fmt.Println(" Complete!")
	*/

}

func testFunc(file *os.File, f func(int) float64, funcName string, n int) {
	avg := 0.0

	for i:= 0; i < 30 ; i++{
		t:= f(n)
		avg += t

		line := fmt.Sprintf("%s,%d,%.6f\n", funcName, n, t)

		_, err := file.WriteString(line)
		if err != nil {
			fmt.Println("Error:", err)
		}
	
	}

	avg /= 30

	avgLine := fmt.Sprintf("%s,%d,Average,%.6f\n\n", funcName, n, avg)
	_, err := file.WriteString(avgLine)
	if err != nil {
		fmt.Println("Error writing average:", err)
	}

}

func testBlockFunc(f func(int,int) float64, n int, bkSize int) {

	avg := 0.0

	for i:= 0; i < 30; i++ {
		avg += f(n,bkSize)
	}

	avg /= 3.0

	// put here code for csv

}
